python_modules/dagster/dagster/_core/asset_graph_view/asset_graph_view.py

```diff
@@ -0,0 +1,276 @@
+    * partition mappings and query the instance.
+    * Methods using "materialize" indicate that they fully materialize partition sets
+      These can potentially be very expensive if the underlying partition set has
+      an in-memory representation that involves large time windows. I.e. if
+      the underlying PartitionsSubset in the ValidAssetSubset is a TimeWindowPartitionsSubset
+      Usage of these methods should be avoided if possible if you are potentially
+      dealing with slices with large time-based partition windows.
+    """
+
+    def __init__(
+        self, asset_graph_view: "AssetGraphView", compatible_subset: _AssetSliceCompatibleSubset
+    ):
+        self._asset_graph_view = asset_graph_view
+        self._compatible_subset = compatible_subset
+
+    def convert_to_valid_asset_subset(self) -> ValidAssetSubset:
+        return self._compatible_subset
+
+    def compute_partition_keys(self) -> AbstractSet[PartitionKey]:
+        return {ap.partition_key for ap in self._compatible_subset.asset_partitions}
+
+    @property
+    def asset_key(self) -> AssetKey:
+        return self._compatible_subset.asset_key
+
+    @property
+    def parent_keys(self) -> AbstractSet[AssetKey]:
+        return self._asset_graph_view.asset_graph.get_parents(self.asset_key)
+
+    @property
+    def _partitions_def(self) -> Optional["PartitionsDefinition"]:
+        return self._asset_graph_view.asset_graph.get_partitions_def(self.asset_key)
+
+    @cached_method
+    def compute_parent_slice(self, parent_asset_key: AssetKey) -> "AssetSlice":
+        return self._asset_graph_view.compute_parent_asset_slice(parent_asset_key, self)
+
+    @cached_method
+    def compute_child_slice(self, child_asset_key: AssetKey) -> "AssetSlice":
+        return self._asset_graph_view.compute_child_asset_slice(child_asset_key, self)
+
+    @cached_method
+    def compute_parent_slices(self) -> Mapping[AssetKey, "AssetSlice"]:
+        return {
+            parent_asset_key: self.compute_parent_slice(parent_asset_key)
+            for parent_asset_key in self._asset_graph_view.asset_graph.get_parents(self.asset_key)
+        }
+
+    @cached_method
+    def compute_child_slices(self) -> Mapping[AssetKey, "AssetSlice"]:
+        return {
+            child_asset_key: self.compute_child_slice(child_asset_key)
+            for child_asset_key in self._asset_graph_view.asset_graph.get_children(self.asset_key)
+        }
+
+    def only_partition_keys(self, partition_keys: AbstractSet[PartitionKey]) -> "AssetSlice":
+        """Return a new AssetSlice with only the given partition keys if they are in the slice."""
+        return _slice_from_subset(
+            self._asset_graph_view,
+            AssetSubset.from_asset_partitions_set(
+                self.asset_key,
+                self._partitions_def,
+                {AssetPartition(self.asset_key, partition_key) for partition_key in partition_keys},
+            )
+            & self._compatible_subset,
```

Comment:
nit: this should be `self._compatible_subset & AssetSubset.from_partitions_set(...)` for perf. This method will often be called on an AssetSlice which contains an AllPartitionsSubset, which has a "free" implementation of `__and__` (`AllPartitionsSubset & X = X`), but if we do `X & AllPartitionsSubset`, the default `__and__` implementation will force us to construct the keys in the AllPartitionsSubset.

Reply:
Let's specialize the base class operators instead so we don't have to think about it. https://github.com/dagster-io/dagster/pull/20354